### ISSUE TITLE
🚦 Traffic Sign Recognition using HOG + SVM

### ISSUE THREAD

**Issue by 24CS059Aemi** [Niketkumardheeryan/ML-CaPsule#1511]
<html>
<body>
<h1 data-section-id="18xpys" data-start="254" data-end="269">📌 Overview</h1>
<p data-start="270" data-end="479">Traffic signs are essential for safe driving and automated navigation systems.<br data-start="348" data-end="351">
This project builds a <strong data-start="373" data-end="408">Traffic Sign Recognition System</strong> using classical Machine Learning.<br data-start="442" data-end="445">
Instead of deep learning, it uses:</p>
<ul data-start="481" data-end="624">
<li data-section-id="1988q61" data-start="481" data-end="561">
<strong data-start="483" data-end="524">HOG (Histogram of Oriented Gradients)</strong> to extract edge and shape features
</li>
<li data-section-id="1ro6hh8" data-start="562" data-end="624">
<strong data-start="564" data-end="596">SVM (Support Vector Machine)</strong> to classify traffic signs
</li>
</ul>
<p data-start="626" data-end="722">This approach is lightweight, explainable, and works well even with limited computational power.</p>
<hr data-start="724" data-end="727">
<h1 data-section-id="1p4t2ly" data-start="729" data-end="749">🎯 Project Goals</h1>
<ul data-start="750" data-end="1127">
<li data-section-id="qyndik" data-start="750" data-end="804">
Load and preprocess the GTSRB traffic sign dataset
</li>
<li data-section-id="1l60yzh" data-start="805" data-end="877">
Convert images to grayscale, resize them, and normalize pixel values
</li>
<li data-section-id="1dx8zgj" data-start="878" data-end="922">
Extract <strong data-start="888" data-end="904">HOG features</strong> for every image
</li>
<li data-section-id="lat2xg" data-start="923" data-end="965">
Train a <strong data-start="933" data-end="963">multi-class SVM classifier</strong>
</li>
<li data-section-id="1sr3dos" data-start="966" data-end="1032">
Evaluate accuracy, confusion matrix, and classification report
</li>
<li data-section-id="eihxb9" data-start="1033" data-end="1081">
Test the model on unseen traffic sign images
</li>
<li data-section-id="6z34ii" data-start="1082" data-end="1127">
Visualize predictions and compare results
</li>
</ul>
<hr data-start="1129" data-end="1132">
<h1 data-section-id="1ripjy8" data-start="1134" data-end="1152">🛠️ Tech Stack</h1>
<div class="TyagGW_tableContainer"><div tabindex="-1" class="group TyagGW_tableWrapper flex flex-col-reverse w-fit">
Tool | Purpose
-- | --
Python 3.x | Core programming language
Scikit-image | HOG feature extraction
Scikit-learn | SVM model + evaluation metrics
OpenCV | Image loading & preprocessing
NumPy | Data handling
Matplotlib / Seaborn | Plotting & visualization

</div></div>
<hr data-start="1448" data-end="1451">
<h1 data-section-id="ukfucl" data-start="1453" data-end="1472">📂 Deliverables</h1>
<ul data-start="1473" data-end="1751">
<li data-section-id="oseopu" data-start="1473" data-end="1547">
<strong data-start="1475" data-end="1509">traffic_sign_recognition.ipynb</strong> — complete, well-explained notebook
</li>
<li data-section-id="4lql57" data-start="1548" data-end="1594">
<strong data-start="1550" data-end="1567">svm_model.pkl</strong> — trained SVM classifier
</li>
<li data-section-id="13zdabk" data-start="1595" data-end="1651">
<strong data-start="1597" data-end="1617">hog_features.npy</strong> — saved HOG features (optional)
</li>
<li data-section-id="lq4zgs" data-start="1652" data-end="1700">
<strong data-start="1654" data-end="1678">confusion_matrix.png</strong> — evaluation output
</li>
<li data-section-id="1v1odkq" data-start="1701" data-end="1751">
<strong data-start="1703" data-end="1716">README.md</strong> — project overview + setup steps
</li>
</ul>
<hr data-start="1753" data-end="1756">
<h1 data-section-id="kzpmsj" data-start="1758" data-end="1772">📊 Dataset</h1>
<p data-start="1773" data-end="1925">We use the <strong data-start="1784" data-end="1837">German Traffic Sign Recognition Benchmark (GTSRB)</strong> dataset.<br data-start="1846" data-end="1849">
It contains <strong data-start="1861" data-end="1875">43 classes</strong> of traffic signs and thousands of labeled images.</p>
<p data-start="1927" data-end="1993">Dataset link:<br data-start="1940" data-end="1943">
👉 <a data-start="1946" data-end="1993" rel="noopener" target="_new" class="decorated-link" href="https://benchmark.ini.rub.de/gtsrb_dataset.html">https://benchmark.ini.rub.de/gtsrb_dataset.html</a></p>
<p data-start="1995" data-end="2010">You may either:</p>
<ul data-start="2012" data-end="2091">
<li data-section-id="17jac02" data-start="2012" data-end="2040">
use the full dataset, or
</li>
<li data-section-id="1xdokx4" data-start="2041" data-end="2091">
work with a smaller subset for faster training
</li>
</ul>
<hr data-start="2093" data-end="2096">
<h1 data-section-id="1759rv4" data-start="2098" data-end="2121">🔍 Project Workflow</h1>
<h3 data-section-id="cxkxvz" data-start="2123" data-end="2148"><span role="text"><strong data-start="2127" data-end="2146">1. Data Loading</strong></span></h3>
<p data-start="2149" data-end="2190">Load the GTSRB dataset (images + labels).</p>
<h3 data-section-id="12hyz12" data-start="2192" data-end="2218"><span role="text"><strong data-start="2196" data-end="2216">2. Preprocessing</strong></span></h3>
<ul data-start="2219" data-end="2315">
<li data-section-id="xmswrk" data-start="2219" data-end="2263">
Resize each image (e.g., 64×64 or 32×32)
</li>
<li data-section-id="uwptox" data-start="2264" data-end="2288">
Convert to grayscale
</li>
<li data-section-id="w6u7rx" data-start="2289" data-end="2315">
Normalize pixel values
</li>
</ul>
<h3 data-section-id="1x2wiqy" data-start="2317" data-end="2352"><span role="text"><strong data-start="2321" data-end="2350">3. HOG Feature Extraction</strong></span></h3>
<p data-start="2353" data-end="2418">Extract Histogram of Oriented Gradients features for every image.</p>
<h3 data-section-id="1s49w2c" data-start="2420" data-end="2445"><span role="text"><strong data-start="2424" data-end="2443">4. SVM Training</strong></span></h3>
<p data-start="2446" data-end="2510">Train a <strong data-start="2454" data-end="2480">Support Vector Machine</strong> classifier using HOG vectors.</p>
<h3 data-section-id="x99zd7" data-start="2512" data-end="2535"><span role="text"><strong data-start="2516" data-end="2533">5. Evaluation</strong></span></h3>
<p data-start="2536" data-end="2547">Generate:</p>
<ul data-start="2548" data-end="2613">
<li data-section-id="2axfj1" data-start="2548" data-end="2566">
Accuracy score
</li>
<li data-section-id="1snwfll" data-start="2567" data-end="2587">
Confusion matrix
</li>
<li data-section-id="1hk1ki8" data-start="2588" data-end="2613">
Classification report
</li>
</ul>
<h3 data-section-id="auku7f" data-start="2615" data-end="2638"><span role="text"><strong data-start="2619" data-end="2636">6. Prediction</strong></span></h3>
<p data-start="2639" data-end="2705">Test the model on new traffic sign images and display predictions.</p>
</body>
</html>
